### PR TITLE
Feature/1 improve errors and tooltips

### LINKED ISSUE #1: Improve descriptions & error messages

### PR DESCRIPTION
**Pull request by sklein94**
Resolve #1

### CODE REVIEW

**Comment by JBartscher** [comment]
# Summary

Added tooltips for configuration of widgets. It was mostly reusable code from the github widgets.
However there was a problem with the "required" attribute in the 'SCM Markdown Content
'-widget. If this attribute was attached to the <selects> the form could no longer be submitted. Therefore, the "required" attribute was only attached to the markdown-path field, but now the markdown-path is only displayed if a branch is available (per angular condition). Ergo, a repository has been selected.
The tooltip message was also extended a bit with more tips on how to fill the fields.

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /build/
 /components/
 .tmp/
+.idea/
```
Comment (Phil-Ah) [inline .gitignore:6]: Please do not push the .idea -folder.

File: dist/adf-widget-scm.js [hunk 2]
```diff
@@ -183,17 +183,22 @@ angular.module('adf.widget.scm', ['adf.provider', 'chart.js', 'ngSanitize', 'btf
...
+$templateCache.put("{widgetsPath}/scm/src/edit/edit.html","<form role=form><div class=form-group><label for=repository>Repository</label> <span class=\"grey glyphicon glyphicon-info-sign ng-scope\" tabindex=0 ng-focus=\"repositoryTooltipIsOpen = !repositoryTooltipIsOpen\" tooltip-trigger=outsideClick tooltip-is-open=repositoryTooltipIsOpen ng-mouseenter=\"repositoryTooltipIsOpen = !repositoryTooltipIsOpen\" uib-tooltip-html=repositoryTooltip ng-mouseover=\"tooltipIsOpenRepo = !tooltipIsOpenRepo\" tooltip-placement=top aria-label=\"A valid URL to your Jenkins server must be stored here\"></span><select name=repository id=repository class=form-control ng-model=config.repository required><option ng-repeat=\"repository in vm.repositories | orderBy: \'name\'\" value={{repository.id}}>{{repository.name}}</option></select></div></form>");
```
Comment (Phil-Ah) [inline dist/adf-widget-scm.js]: Is the URL of the Jenkins server the right default message?

File: dist/adf-widget-scm.js [hunk 3]
```diff
@@ -183,17 +183,22 @@ angular.module('adf.widget.scm', ['adf.provider', 'chart.js', 'ngSanitize', 'btf
...
...hes><label for=branch>Branch</label> <span class=\"grey glyphicon glyphicon-info-sign ng-scope\" tabindex=0 ng-focus=\"branchTooltipIsOpen = !branchTooltipIsOpen\" tooltip-trigger=outsideClick uib-tooltip-html=branchTooltip ng-mouseover=\"branchTooltipIsOpen = !branchTooltipIsOpen\" tooltip-placement=top tooltip-is-open=branchTooltipIsOpen aria-label=\"A valid URL to your Jenkins server must be stored here\"></span><select name=branch id=branch class=form-control ng-model=config.branch><option ng-repeat=\"branch in vm.branches| orderBy: \'name\'\" value={{branch.id}}>{{branch.name}}</option></select></p><p ng-if=vm.branches><label for=path>Path to Markdown File</label> <span class=\"grey glyphicon glyphicon-info-sign ng-scope\" tabindex=0 ng-focus=\"pathToMarkdownTooltipIsOpen = !pathToMarkdownTooltipIsOpen\" tooltip-trigger=outsideClick uib-tooltip-html=pathToMarkdownTooltip ng-mouseover=\"pathToMarkdownTooltipIsOpen = !pathToMarkdownTooltipIsOpen\" tooltip-is-open=pathToMarkdownTooltipIsOpen tooltip-placement=top aria-label=\"A valid path to a markdown file must be stored here\"></span> <input type=text class=form-control id=path ng-model=config.path required></p></div></form>");
```
Comment (Phil-Ah) [inline dist/adf-widget-scm.js]: Is the URL of the Jenkins server the right default message?
Please refer to voice & tone when writing messages. “A valid path to a markdown file must be stored here” might be rewritten to “Enter the path to a markdown file”.
Add information if the format of the path has to follow a special format (no leading / for example)

File: dist/adf-widget-scm.js [hunk 4]
```diff
@@ -183,17 +183,22 @@ angular.module('adf.widget.scm', ['adf.provider', 'chart.js', 'ngSanitize', 'btf
...
+$templateCache.put("{widgetsPath}/scm/src/markdownPreview/view.html","<style>\n  div.markdownContent{\n    overflow: auto;\n    width: 100%;\n  }\n</style><div class=\"alert alert-info\" ng-if=\"vm.fileContent === null\">Please configure a specific file</div><div ng-if=vm.fileContent btf-markdown=vm.fileContent class=markdownContent></div><div class=\"alert alert-danger\" ng-if=\"vm.fileContent.status == 500 || vm.fileContent.status == 404\"><b>Error {{vm.fileContent.status}}</b> The Markdown file was not found. This may be because the file does not exist or the SCM Manager is not available at the moment.</div>");}]);
```
Comment (Phil-Ah) [inline dist/adf-widget-scm.js]: “Configure a specific file”

File: dist/adf-widget-scm.js [hunk 5]
```diff
@@ -183,17 +183,22 @@ angular.module('adf.widget.scm', ['adf.provider', 'chart.js', 'ngSanitize', 'btf
...
 angular.module('adf.widget.scm')
-  .controller('MarkdownPreviewEditController', ["repositories", "SCM", function (repositories, SCM) {
+  .controller('MarkdownPreviewEditController', ["repositories", "SCM", "$scope", "$sce", function (repositories, SCM, $scope, $sce) {
     var vm = this;
     vm.repositories = repositories;
 
+    // tooltips
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository which stores the desired Markdown.');
+    $scope.branchTooltip = $sce.trustAsHtml('Select the branch on which the desired Markdown file can be found.');
+    $scope.pathToMarkdownTooltip = $sce.trustAsHtml('The path where the Markdown file can be found within the repository (e.g. <b>docs/README.md</b>). Note that the path is case sensitive. For files in the root dir of the repository add <b>NO</b> leading slash.');
```
Comment (Phil-Ah) [inline dist/adf-widget-scm.js]: “Enter the path to the Markdown file with a leading slash. The path is case sensitive. Files in the root dir start <b>without</b> a leading slash.

File: dist/adf-widget-scm.js [hunk 6]
```diff
@@ -288,9 +293,11 @@ angular.module('adf.widget.scm')
 
 
 angular.module('adf.widget.scm')
-  .controller('ScmEditController', ["repositories", function(repositories){
+  .controller('ScmEditController', ["repositories", "$scope", "$sce", function(repositories, $scope, $sce){
     var vm = this;
 
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository which is used to provide data.');
```
Comment (Phil-Ah) [inline dist/adf-widget-scm.js]: “Select a repository”

File: src/edit/edit.html [hunk 7]
```diff
@@ -1,7 +1,8 @@
 <form role="form">
   <div class="form-group">
     <label for="repository">Repository</label>
-    <select name="repository" id="repository" class="form-control" ng-model="config.repository">
+    <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0" ng-focus="repositoryTooltipIsOpen = !repositoryTooltipIsOpen" tooltip-trigger="outsideClick" tooltip-is-open="repositoryTooltipIsOpen" ng-mouseenter="repositoryTooltipIsOpen = !repositoryTooltipIsOpen" uib-tooltip-html="repositoryTooltip" tooltip-trigger="'outsideClick'" tooltip-is-open="tooltipIsOpenRepo" ng-mouseover="tooltipIsOpenRepo = !tooltipIsOpenRepo" tooltip-placement="top" aria-label="A valid URL to your Jenkins server must be stored here"></span>
```
Comment (Phil-Ah) [inline src/edit/edit.html]: Is the URL of the Jenkins server the right default message?

File: src/edit/simpleEditController.js [hunk 8]
```diff
@@ -25,8 +25,10 @@
 'use strict';
 
 angular.module('adf.widget.scm')
-  .controller('ScmEditController', function(repositories){
+  .controller('ScmEditController', function(repositories, $scope, $sce){
     var vm = this;
 
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository which is used to provide data.');
```
Comment (Phil-Ah) [inline src/edit/simpleEditController.js]: “Select a repository”

File: src/markdownPreview/edit.html [hunk 9]
```diff
@@ -1,6 +1,13 @@
...
   <div class="form-group">
-    <p><label for="repository">Repository</label>
+    <p>
+      <label for="repository">Repository</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
+            ng-focus="repositoryTooltipIsOpen = !repositoryTooltipIsOpen" tooltip-trigger="outsideClick"
+            tooltip-is-open="repositoryTooltipIsOpen" ng-mouseenter="repositoryTooltipIsOpen = !repositoryTooltipIsOpen"
+            uib-tooltip-html="repositoryTooltip" tooltip-trigger="'outsideClick'" tooltip-is-open="tooltipIsOpenRepo"
+            ng-mouseover="tooltipIsOpenRepo = !tooltipIsOpenRepo" tooltip-placement="top"
+            aria-label="A valid URL to your Jenkins server must be stored here"></span>
```
Comment (Phil-Ah) [inline src/markdownPreview/edit.html]: Is the URL of the Jenkins server the right default message?

File: src/markdownPreview/edit.html [hunk 10]
```diff
@@ -10,15 +17,30 @@
...
 
-    <p ng-if="vm.branches"><label for="branch">Branch</label>
+    <p ng-if="vm.branches">
+      <label for="branch">Branch</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
+            ng-focus="branchTooltipIsOpen = !branchTooltipIsOpen" tooltip-trigger="outsideClick"
+            uib-tooltip-html="branchTooltip" tooltip-trigger="'outsideClick'"
+            ng-mouseover="branchTooltipIsOpen = !branchTooltipIsOpen" tooltip-placement="top"
+            tooltip-is-open="branchTooltipIsOpen"
+            aria-label="A valid URL to your Jenkins server must be stored here"></span>
```
Comment (Phil-Ah) [inline src/markdownPreview/edit.html]: Is the URL of the Jenkins server the right default message?

File: src/markdownPreview/markdownPreviewEdit.js [hunk 11]
```diff
@@ -1,10 +1,15 @@
 'use strict';
 
 angular.module('adf.widget.scm')
-  .controller('MarkdownPreviewEditController', function (repositories, SCM) {
+  .controller('MarkdownPreviewEditController', function (repositories, SCM, $scope, $sce) {
     var vm = this;
     vm.repositories = repositories;
 
+    // tooltips
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository which stores the desired Markdown.');
```
Comment (Phil-Ah) [inline src/markdownPreview/markdownPreviewEdit.js]: Please refer to voice & tone when writing messages. “A valid path to a markdown file must be stored here” might be rewritten to “Enter the path to a markdown file”.
Add information if the format of the path has to follow a special format (no leading / for example)

File: src/markdownPreview/edit.html [hunk 12]
```diff
@@ -10,15 +17,30 @@
...
-    <label for="path">Path to Markdown File</label>
-    <input type="text" class="form-control" id="path" ng-model="config.path">
+    <p ng-if="vm.branches">
+      <label for="path">Path to Markdown File</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
+            ng-focus="pathToMarkdownTooltipIsOpen = !pathToMarkdownTooltipIsOpen" tooltip-trigger="outsideClick"
+            uib-tooltip-html="pathToMarkdownTooltip" tooltip-trigger="'outsideClick'"
+            ng-mouseover="pathToMarkdownTooltipIsOpen = !pathToMarkdownTooltipIsOpen"
+            tooltip-is-open="pathToMarkdownTooltipIsOpen"
+            tooltip-placement="top" aria-label="A valid path to a markdown file must be stored here"></span>
```
Comment (Phil-Ah) [inline src/markdownPreview/edit.html]: Please refer to voice & tone when writing messages. “A valid path to a markdown file must be stored here” might be rewritten to “Enter the path to a markdown file”.
Add information if the format of the path has to follow a special format (no leading / for example)

File: src/markdownPreview/markdownPreviewEdit.js [hunk 13]
```diff
@@ -1,10 +1,15 @@
...
+    $scope.branchTooltip = $sce.trustAsHtml('Select the branch on which the desired Markdown file can be found.');
```
Comment (Phil-Ah) [inline src/markdownPreview/markdownPreviewEdit.js]: Refer to prior suggestions for wording.

**Review by Phil-Ah — CHANGES_REQUESTED** [review]
Good work adding all the tooltips. Some may be expressed more clearly. See comments below.

File: src/edit/simpleEditController.js [hunk 14]
```diff
@@ -25,8 +25,10 @@
 'use strict';
 
 angular.module('adf.widget.scm')
-  .controller('ScmEditController', function(repositories){
+  .controller('ScmEditController', function(repositories, $scope, $sce){
     var vm = this;
 
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository.');
```
Comment (mbehlendorf) [inline src/edit/simpleEditController.js:31]: Maybe use parseAsHtml?

File: src/markdownPreview/edit.html [hunk 15]
```diff
@@ -1,6 +1,13 @@
 <form role="form">
   <div class="form-group">
-    <p><label for="repository">Repository</label>
+    <p>
+      <label for="repository">Repository</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
```
Comment (mbehlendorf) [inline src/markdownPreview/edit.html]: Why using 2 variables here? `tooltipIsOpenRepo` and `repositoryTooltipIsOpen`
Some attributes are defined 2 times: e.g. `tooltip-is-open`

File: src/markdownPreview/edit.html [hunk 16]
```diff
@@ -10,15 +17,30 @@
       </select>
     </p>
 
-    <p ng-if="vm.branches"><label for="branch">Branch</label>
+    <p ng-if="vm.branches">
+      <label for="branch">Branch</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
```
Comment (mbehlendorf) [inline src/markdownPreview/edit.html]: see above

File: src/markdownPreview/edit.html [hunk 17]
```diff
@@ -10,15 +17,30 @@
...
         <option ng-repeat="branch in vm.branches| orderBy: 'name'" value="{{branch.id}}">
           {{branch.name}}
         </option>
       </select>
     </p>
-    <label for="path">Path to Markdown File</label>
-    <input type="text" class="form-control" id="path" ng-model="config.path">
+    <p ng-if="vm.branches">
+      <label for="path">Path to Markdown File</label>
+      <span class="grey glyphicon glyphicon-info-sign ng-scope" tabindex="0"
```
Comment (mbehlendorf) [inline src/markdownPreview/edit.html]: see above

File: src/markdownPreview/markdownPreviewEdit.js [hunk 18]
```diff
@@ -1,10 +1,15 @@
 'use strict';
 
 angular.module('adf.widget.scm')
-  .controller('MarkdownPreviewEditController', function (repositories, SCM) {
+  .controller('MarkdownPreviewEditController', function (repositories, SCM, $scope, $sce) {
     var vm = this;
     vm.repositories = repositories;
 
+    // tooltips
+    $scope.repositoryTooltip = $sce.trustAsHtml('Select the repository. If none is shown, check your scm configuration.');
```
Comment (mbehlendorf) [inline src/markdownPreview/markdownPreviewEdit.js:9]: see above

**Pull request closed without merging**